File: 백준/Unrated/20005. 보스몬스터 전리품/보스몬스터 전리품.py
import sys
from collections import deque

input = sys.stdin.readline


def solution(M, N, P, board, idDps, HP):
    EMPTY, BLOCK, BOSS = ".", "X", "B"
    ID, DPS, INIT_POSITION = 0, 1, 2
    INDEX, DISTANCE = 0, 1
    direction = [[0, 1], [1, 0], [0, -1], [-1, 0]]

    def outOfBound(r, c):
        return r < 0 or c < 0 or M <= r or N <= c

    def bfs(startR, startC):
        head = 0
        queue = deque([[startR, startC]])
        bossR, bossC = boss
        distance = [[-1] * N for _ in range(M)]
        distance[startR][startC] = 0

        while queue:
            r, c = queue.popleft()

            for dr, dc in direction:
                nr, nc = r + dr, c + dc
                if outOfBound(nr, nc):
                    continue
                if board[r][c] == BLOCK:
                    continue
                if distance[nr][nc] != -1:
                    continue

                distance[nr][nc] = distance[r][c] + 1

                if nr == bossR and nc == bossC:
                    return distance[nr][nc]
                queue.append([nr, nc])

        return -1

    boss = None
    player = []
    playerMap = {}

    distanceToBoss = []
    for i in range(len(idDps)):
        id, dps = idDps[i].split()
        player.append([id, int(dps), None])
        playerMap[id] = i

    for r in range(M):
        for c in range(N):
            if board[r][c] == BOSS:
                boss = [r, c]
            elif board[r][c] not in (EMPTY, BLOCK):
                index = playerMap[board[r][c]]
                player[index][INIT_POSITION] = [r, c]

    for index, info in enumerate(player):
        r, c = info[INIT_POSITION]
        distance = bfs(r, c)
        if distance == -1:
            continue
        distanceToBoss.append([index, distance])

    t = 0
    curDps = 0
    distanceToBoss.sort(key=lambda x: x[DISTANCE], reverse=True)

    while HP > 0 and distanceToBoss:
        index, distance = distanceToBoss.pop()
        dps = player[index][DPS]
        if t < distance:
            dt = distance - t
            HP -= curDps * dt
            t = distance

            if HP < 0:
                distanceToBoss.append([index, distance])
                break
        curDps += dps

    return P - len(distanceToBoss)


M, N, P = map(int, input().split())
board = [list(input().rstrip()) for _ in range(M)]
idDps = [input().rstrip() for _ in range(P)]
HP = int(input().rstrip())

print(solution(M, N, P, board, idDps, HP))

File: 백준/Gold/20005. 보스몬스터 전리품/보스몬스터 전리품.js
const fs = require("fs");
const path = process.platform === "linux" ? "/dev/stdin" : "예제.txt";
const input = fs.readFileSync(path).toString().trim().split("\n");

function solution(M, N, P, board, player, hp) {
  const [BOSS, WALL] = ["B", "X"];
  const direction = [
    [0, 1],
    [1, 0],
    [0, -1],
    [-1, 0],
  ];

  let boss;
  const playerMap = new Map(player);
  const outOfBound = (r, c) => r < 0 || c < 0 || M <= r || N <= c;
  const play = (bossR, bossC) => {
    let t = 0;
    let accDps = 0;
    let nPlayer = 0;

    let head = 0;
    const queue = [];
    const distance = Array.from(Array(M), () => Array(N).fill(-1));

    distance[bossR][bossC] = 0;
    queue.push([bossR, bossC]);

    while (queue.length - head) {
      const [r, c] = queue[head++];

      if (playerMap.has(board[r][c])) {
        if (t < distance[r][c]) {
          hp -= (distance[r][c] - t) * accDps;
          t = distance[r][c];
        }
        if (hp <= 0) return nPlayer;

        accDps += playerMap.get(board[r][c]);
        nPlayer += 1;
      }

      for (const [dr, dc] of direction) {
        const [nr, nc] = [r + dr, c + dc];
        if (outOfBound(nr, nc)) continue;
        if (board[nr][nc] === WALL) continue;
        if (distance[nr][nc] >= 0) continue;

        distance[nr][nc] = distance[r][c] + 1;
        queue.push([nr, nc]);
      }
    }

    return nPlayer;
  };

  for (let index = 0; index < N * M; index++) {
    const [r, c] = [Math.floor(index / N), index % N];
    if (board[r][c] === BOSS) {
      boss = [r, c];
      break;
    }
  }

  return play(...boss);
}

const [M, N, P] = input.shift().split(" ").map(Number);
const board = input.splice(0, M);
const player = input.splice(0, P).map((row) => {
  const [id, dps] = row.split(" ");
  return [id, Number(dps)];
});
const hp = Number(input[0]);
console.log(solution(M, N, P, board, player, hp));
